src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/AsyncExceptionHandlerRewriter.cs

```diff
@@ -518,11 +519,18 @@ private BoundNode VisitCatchBlock(BoundCatchBlock node, AwaitCatchFrame parentAw
                 var origCurrentAwaitCatchFrame = _currentAwaitCatchFrame;
                 _currentAwaitCatchFrame = parentAwaitCatchFrame;
 
+                var origInCatchWithoutAwaits = _inCatchWithoutAwaits;
+                _inCatchWithoutAwaits = true;
+
                 var result = base.VisitCatchBlock(node);
                 _currentAwaitCatchFrame = origCurrentAwaitCatchFrame;
+                _inCatchWithoutAwaits = origInCatchWithoutAwaits;
                 return result;
             }
 
+            // We cannot get here from a catch without awaits.
+            Debug.Assert(!_inCatchWithoutAwaits);
```

Comment:
>Debug.Assert(!_inCatchWithoutAwaits); [](http://example.com/codeflow?start=12&length=37)

Are we certain this is the case? What if a catch without awaits contains another try/catch that uses awaits? The outer catch doesn't use await directly, but the nested one does. Is this assert going to fail when we reach that nested catch with awaits? Are we testing a scenario like that?

Reply:
It looks like `NestedRethrow_02` covers this scenario

Reply:
Yes, the "await analysis" for the outer `catch` visits the whole block including the nested `catch`es, I think, so it will report any outer `catch` to contain `await`s whenever any of its nested `catch`es contains an `await`. Hence we cannot have a `catch` without `await`s and nested inside it a `catch` with `await`s.